### PR TITLE
ASIM Registry schema parser with its sample and test data for SentinelOne

### PR DESCRIPTION
**Pull request by jayeshprajapaticrest**
Required items, please complete
   
   Change(s):
   - Added ASIM Registry  schema parser for SentinelOne based on API Data

   Reason for Change(s):
   - Initial version ASIM Registry  Parser

   Version Updated:
   - Initial version

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

### CODE REVIEW

**Comment by v-rusraut** [comment]
Hi @jayeshprajapaticrest,
Thank you for raising Pull Request with us! We will review it internally and get back to you by shortly.

**Comment by vakohl** [comment]
ETA: for review by 18th Aug

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      SentinelOne_CL
+      | where event_name_s == "Alerts."
+          and alertInfo_eventType_s in ("REGVALUEMODIFIED", "REGVALUECREATE", "REGKEYCREATE", "REGKEYDELETE", "REGVALUEDELETE", "REGKEYRENAME")
+      | lookup EventTypeLookup on alertInfo_eventType_s
+      | lookup RegistryPreviousValueTypeLookup on alertInfo_registryOldValueType_s
+      | invoke _ASIM_ResolveDvcFQDN('agentDetectionInfo_name_s')
+      | extend RegistryKeyPrefix = tostring(split(alertInfo_registryKeyPath_s, @'\')[0])
+      | lookup RegistryKeyPrefixLookup on RegistryKeyPrefix
+      | extend RegistryKey = replace_string(alertInfo_registryKeyPath_s, RegistryKeyPrefix, RegistryKeyNormalizedPrefix)
+      | extend RegistryValue = iff(alertInfo_eventType_s in ("REGVALUEMODIFIED", "REGVALUECREATE", "REGVALUEDELETE") , tostring(split(alertInfo_registryKeyPath_s, @'\')[-1]), "")
```
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml:64]: many rows are missing Other Registry fields like value, type,  even though we have RegistryKey is present
image

File: Parsers/ASimRegistryEvent/test/SentinelOne_vimRegistryEvent_SchemaTest.csv [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿Result
+"(1) Warning: Missing recommended field [ActorUserId]"
```
Comment (vakohl) [inline Parsers/ASimRegistryEvent/test/SentinelOne_vimRegistryEvent_SchemaTest.csv:2]: any id coming for user from source?

**Review by vakohl — CHANGES_REQUESTED** [review]

**Comment by v-sudkharat** [comment]
Hi @jayeshprajapaticrest, hope you are doing well. Just wanted to check if you got a chance to look at the suggestions shared. Please feel free to reach out to us for any queries or support. Thanks!

**Comment by v-sudkharat** [comment]
Hi @jayeshprajapaticrest, we are waiting for your response. Thanks!

File: Parsers/ASimRegistryEvent/test/SentinelOne_vimRegistryEvent_DataTest.csv [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿Result
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 159 records (100.0%) for field [EventProduct] of type [Enumerated]: [""SentinelOne""] (Schema:RegistryEvent)"
+"(0) Error: 2 invalid value(s) (up to 10 listed) in 130 records (81.76%) for field [EventType] of type [Enumerated]: [""RegistryValueSet"",""RegistryValueDeleted""] (Schema:RegistryEvent)"
+"(2) Info: Empty value in 159 records (100.0%)  in recommended field [EventUid] (Schema:RegistryEvent)"
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/test/SentinelOne_vimRegistryEvent_DataTest.csv]: Done

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml [hunk 4]
```diff
@@ -0,0 +1,118 @@
+Parser:
+  Title: Registry Event ASIM Parser for SentinelOne
+  Version: '0.1'
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml]: Done

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+          DvcId = agentDetectionInfo_uuid_g,
+          DvcOs = agentDetectionInfo_osName_s,
+          DvcOsVersion = agentDetectionInfo_osRevision_s,
+          EventOriginalType = alertInfo_eventType_s,
+          ParentProcessName = sourceParentProcessInfo_name_s,
+          RegistryValueData = alertInfo_registryValue_s,
+          EventOriginalUid = alertInfo_dvEventId_s
+      | extend
+          EventCount = int(1),
+          EventProduct = "SentinelOne",
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml:95]: Done.

[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml:95]: Done.

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+                                 alertInfo_registryValue_s startswith "0x" and strlen(alertInfo_registryValue_s) > 10, "Reg_QWord",
+                                 alertInfo_registryValue_s matches regex '^[A-Fa-f0-9]+$', "Reg_Binary",
+                                 ""
+                             )
+      | extend RegistryValueType = iff(alertInfo_eventType_s in ("REGVALUEMODIFIED", "REGVALUECREATE") and isempty(RegistryValueType), "Reg_Sz", RegistryValueType)
+      | project-rename
+          ActingProcessId = sourceProcessInfo_pid_s,
+          ActorUsername = sourceProcessInfo_user_s,
+          EventStartTime= sourceProcessInfo_pidStarttime_t,
+          EventUid = _ResourceId,
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml]: Done.

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+          ParentProcessId = sourceParentProcessInfo_pid_s,
+          ActingProcessName = sourceProcessInfo_name_s,
+          DvcId = agentDetectionInfo_uuid_g,
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml:86]: Done.

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+          DvcOs = agentDetectionInfo_osName_s,
+          DvcOsVersion = agentDetectionInfo_osRevision_s,
+          EventOriginalType = alertInfo_eventType_s,
+          ParentProcessName = sourceParentProcessInfo_name_s,
+          RegistryValueData = alertInfo_registryValue_s,
+          EventOriginalUid = alertInfo_dvEventId_s
+      | extend
+          EventCount = int(1),
+          EventProduct = "SentinelOne",
+          EventResult = "Success",
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml:97]: Done.
As Discussed We are setting DvcAction="Allowed" as EventResult is always "Success".

File: Parsers/ASimRegistryEvent/test/SentinelOne_vimRegistryEvent_SchemaTest.csv (hunk 2 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/test/SentinelOne_vimRegistryEvent_SchemaTest.csv:2]: No. For "Alerts" type of data from SentinelOne we are not getting userId for RegistryEvent Data.

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+Normalization:
+  Schema: RegistryEvent
+  Version: '0.1.2'
+References:
+- Title: ASIM Registry Schema
+  Link: https://aka.ms/ASimRegistryEventDoc
+- Title: ASIM
+  Link: https:/aka.ms/AboutASIM
+- Title: SentinelOne documentation 
+  Link: https://<SOneInstanceDomain>.sentinelone.net/api-doc/overview
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml:16]: As discussed the SentinelOne doesn't have the public documentation available for API.
User can refer from their instance URL.

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+      "MULTI_SZ", "Reg_Multi_Sz",
+      "DWORD_BIG_ENDIAN", "Reg_DWord"
+  ];
+  let parser = () { 
+      SentinelOne_CL
+      | where event_name_s == "Alerts."
+          and alertInfo_eventType_s in ("REGVALUEMODIFIED", "REGVALUECREATE", "REGKEYCREATE", "REGKEYDELETE", "REGVALUEDELETE", "REGKEYRENAME")
+      | lookup EventTypeLookup on alertInfo_eventType_s
+      | lookup RegistryPreviousValueTypeLookup on alertInfo_registryOldValueType_s
+      | invoke _ASIM_ResolveDvcFQDN('agentDetectionInfo_name_s')
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml:65]: As discussed we have considered agent-related fields as Device information.

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml:69]: 1. In SentinelOne, RegistryPath contains the RegistryValueName as last value in alertInfo_registryKeyPath_s field when type is "REGVALUEMODIFIED", "REGVALUECREATE", "REGVALUEDELETE" so we have extracted from it.
2.  RegistryValue, RegistryValueData() and RegistryValueType(alertInfo_registryValue_s) fields are available only when type is "REGVALUEMODIFIED", "REGVALUECREATE", "REGVALUEDELETE".

**Comment by v-sudkharat** [comment]
Hi @jayeshprajapaticrest, thank you for accommodating the suggested changes! We will review the Pull Request.

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+      | extend RegistryKeyPrefix = tostring(split(alertInfo_registryKeyPath_s, @'\')[0])
+      | lookup RegistryKeyPrefixLookup on RegistryKeyPrefix
+      | extend RegistryKey = replace_string(alertInfo_registryKeyPath_s, RegistryKeyPrefix, RegistryKeyNormalizedPrefix)
+      | extend RegistryValue = iff(alertInfo_eventType_s in ("REGVALUEMODIFIED", "REGVALUECREATE", "REGVALUEDELETE") , tostring(split(alertInfo_registryKeyPath_s, @'\')[-1]), "")
+      | extend RegistryValueType = case(
+                                 alertInfo_registryValue_s matches regex '^[0-9]+$', "Reg_Dword",
+                                 alertInfo_registryValue_s startswith "0x" and strlen(alertInfo_registryValue_s) <= 10, "Reg_DWord",
+                                 alertInfo_registryValue_s startswith "0x" and strlen(alertInfo_registryValue_s) > 10, "Reg_QWord",
+                                 alertInfo_registryValue_s matches regex '^[A-Fa-f0-9]+$', "Reg_Binary",
```
Comment (vakohl) [inline Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml:74]: Can we avoid regex?

**Review by vakohl — CHANGES_REQUESTED** [review]

**Comment by v-sudkharat** [comment]
Hello @jayeshprajapaticrest, please response to the above comments. Thanks!

**Comment by keyur-cds** [comment]
@v-sudkharat: we're planning to update all SentinelOne-related PRs by tomorrow EOD.

File: Parsers/ASimRegistryEvent/Parsers/imRegistry.yaml [hunk 12]
```diff
@@ -22,4 +22,5 @@ ParserQuery: |
     vimRegistryEventMicrosoft365D,
     vimRegistryEventMicrosoftSysmon,
     // vimRegistryEventMicrosoftSecurityEvents, // Deprecated, now included in vimRegistryEventMicrosoftWindowsEvent. 
-    vimRegistryEventMicrosoftWindowsEvent
\ No newline at end of file
+    vimRegistryEventMicrosoftWindowsEvent,
+    vimRegistryEventSentinelOne
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/imRegistry.yaml:33]: Done. Added below missing properties:
- EquivalentBuiltInParser
- Parsers

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+          DvcOs = agentDetectionInfo_osName_s,
+          DvcOsVersion = agentDetectionInfo_osRevision_s,
+          EventOriginalType = alertInfo_eventType_s,
+          ParentProcessName = sourceParentProcessInfo_name_s,
+          RegistryValueData = alertInfo_registryValue_s,
+          EventOriginalUid = alertInfo_dvEventId_s
+      | extend
+          EventCount = int(1),
+          EventProduct = "SentinelOne",
+          EventVendor = "SentinelOne",
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml:143]: Done

File: Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml (hunk 11 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/vimRegistryEventSentinelOne.yaml]: Registry Value type is not directly available in any specific field. There might be chance to get Registry Value data from "alertInfo_registryValue_s" field that is why we used regex to derive type based on value.

**Comment by niralishah-crest** [comment]
@vakohl we have mapped the inspection fields below in the RegistryEvent parser, which we found from SentinelOne Alerts logs.
RuleName
Rule
ThreatConfidence
ThreatOriginalConfidence

File: Parsers/ASimRegistryEvent/Parsers/imRegistry.yaml [hunk 14]
```diff
@@ -12,14 +12,22 @@ References:
   Link: https://aka.ms/ASimRegistryEventDoc
 - Title: ASIM
   Link: https:/aka.ms/AboutASIM
-
 Description: |
   This ASIM parser supports normalizing Registry Event logs from all supported sources to the ASIM Registry Event normalized schema.
 ParserName: imRegistry
+EquivalentBuiltInParser: _Im_RegistryEvent
```
Comment (vakohl) [inline Parsers/ASimRegistryEvent/Parsers/imRegistry.yaml:18]: Team, please add ASIM file for registry too.

Comment (jayeshprajapaticrest) [reply Parsers/ASimRegistryEvent/Parsers/imRegistry.yaml:18]: @vakohl OK Will add unifying ASIM file.
But do we need to add ASIM parser for SentinelOne Registry schema as well?

File: ASIM/dev/ASimTester/ASimTester.csv [hunk 15]
```diff
@@ -553,7 +553,7 @@ EventProductVersion,string,Optional,Dns,,,
 EventProductVersion,string,Optional,FileEvent,,,
 EventProductVersion,string,Optional,NetworkSession,,,
 EventProductVersion,string,Optional,ProcessEvent,,,
-EventProductVersion,string,Optional,RegistryEvent,,,
+EventProductVersion,string,Optional,RegistryEvent,Enumerated,SentinelOne,
```
Comment (vakohl) [inline ASIM/dev/ASimTester/ASimTester.csv:556]: @jayeshprajapaticrest 
pls correct this

File: Parsers/ASimRegistryEvent/Parsers/imRegistry.yaml (hunk 14 above)
Comment (vakohl) [reply Parsers/ASimRegistryEvent/Parsers/imRegistry.yaml:18]: pls remove the other non-existent parsers from ASIM file. Let me know incase you still get validation error

**Review by vakohl — APPROVED** [review]

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**